test/functional/lua/comment_spec.lua

```diff
@@ -0,0 +1,644 @@
+
+      -- Multiline comments should be computed based on first line 'commentstring'
+      set_lines(lines)
+      toggle_lines(1, 3)
+      local out_lines = get_lines()
+      eq(out_lines[1], '"set background=dark')
+      eq(out_lines[2], '"lua << EOF')
+      eq(out_lines[3], '"print(1)')
+    end)
+
+    it('correctly computes indent', function()
+      toggle_lines(2, 4)
+      eq(get_lines(1, 4), { ' # aa', ' #  aa', ' #' })
+
+      set_lines(example_lines)
+      toggle_lines(4, 4)
+      eq(get_lines(3, 4), { '#' })
+    end)
+
+    it('correctly detects comment/uncomment', function()
+      local validate = function(from, to, ref_lines)
+        set_lines({ '', 'aa', '# aa', '# aa', 'aa', '' })
+        toggle_lines(from, to)
+        eq(get_lines(), ref_lines)
+      end
+
+      -- It should uncomment only if all lines are comments
+      validate(3, 4, { '', 'aa', 'aa', 'aa', 'aa', '' })
+      validate(2, 4, { '', '# aa', '# # aa', '# # aa', 'aa', '' })
+      validate(3, 5, { '', 'aa', '# # aa', '# # aa', '# aa', '' })
+      validate(1, 6, { '#', '# aa', '# # aa', '# # aa', '# aa', '#' })
+    end)
+
+    it('matches comment parts strictly when detecting comment/uncomment', function()
+      local validate = function(from, to, ref_lines)
+        set_lines({ '#aa', '# aa', '#  aa' })
+        toggle_lines(from, to)
+        eq(get_lines(), ref_lines)
+      end
+
+      set_commentstring('#%s')
+      validate(1, 3, { 'aa', ' aa', '  aa' })
+      validate(2, 3, { '#aa', ' aa', '  aa' })
+      validate(3, 3, { '#aa', '# aa', '  aa' })
+
+      set_commentstring('# %s')
+      validate(1, 3, { '# #aa', '# # aa', '# #  aa' })
+      validate(2, 3, { '#aa', 'aa', ' aa' })
+      validate(3, 3, { '#aa', '# aa', ' aa' })
+
+      set_commentstring('#  %s')
+      validate(1, 3, { '#  #aa', '#  # aa', '#  #  aa' })
+      validate(2, 3, { '#aa', '#  # aa', '#  #  aa' })
+      validate(3, 3, { '#aa', '# aa', 'aa' })
+    end)
+
+    it('uncomments on inconsistent indent levels', function()
+      set_lines({ '# aa', ' # aa', '  # aa' })
+      toggle_lines(1, 3)
+      eq(get_lines(), { 'aa', ' aa', '  aa' })
+    end)
+
+    it('respects tabs', function()
+      api.nvim_set_option_value('expandtab', false, { buf = 0 })
+      set_lines({ '\t\taa', '\t\taa' })
+
+      toggle_lines(1, 2)
+      eq(get_lines(), { '\t\t# aa', '\t\t# aa' })
+
+      toggle_lines(1, 2)
+      eq(get_lines(), { '\t\taa', '\t\taa' })
+    end)
+
+    it('works with trailing whitespace', function()
+      -- Without right-hand side
+      set_commentstring('# %s')
+      set_lines({ ' aa', ' aa  ', '  ' })
+      toggle_lines(1, 3)
+      eq(get_lines(), { ' # aa', ' # aa  ', ' #' })
+      toggle_lines(1, 3)
+      eq(get_lines(), { ' aa', ' aa  ', '' })
+
+      -- With right-hand side
+      set_commentstring('%s #')
+      set_lines({ ' aa', ' aa  ', '  ' })
+      toggle_lines(1, 3)
+      eq(get_lines(), { ' aa #', ' aa   #', ' #' })
```

Comment:
how does it work if some lines in the range are commented, but others not? is there a test for that?

Reply:
If at least on line is not commented (blank lines are not taken into account when deciding that), then the whole range is commented. Yes, there are tests for that. It is a "correctly detects comment/uncomment" test.